src/material/list/testing/list-harness.ts

```diff
@@ -0,0 +1,299 @@
+    return this.locatorForAll(this._itemHarness.with(filters))();
+  }
+
+  /**
+   * Gets a list of `ListSection` representing the list items grouped by subheaders. If the list has
+   * no subheaders it is represented as a single `ListSection` with an undefined `heading` property.
+   * @param filters Optional filters used to narrow which list item harnesses are included
+   * @return The list of items matching the given filters, grouped into sections by subheader.
+   */
+  async getItemsBySubheader(filters?: F): Promise<ListSection<C>[]> {
+    const listSections = [];
+    let currentSection: ListSection<C> = {items: []};
+    const itemsAndSubheaders =
+        await this.getItemsSubheadersAndDividers({item: filters, divider: false});
+    for (const itemOrSubheader of itemsAndSubheaders) {
+      if (itemOrSubheader instanceof MatSubheaderHarness) {
+        if (currentSection.heading !== undefined || currentSection.items.length) {
+          listSections.push(currentSection);
+        }
+        currentSection = {heading: await itemOrSubheader.getText(), items: []};
+      } else {
+        currentSection.items.push(itemOrSubheader);
+      }
+    }
+    if (currentSection.heading !== undefined || currentSection.items.length ||
+        !listSections.length) {
+      listSections.push(currentSection);
+    }
+    return listSections;
+  }
+
+  /**
+   * Gets a list of sub-lists representing the list items grouped by dividers. If the list has no
+   * dividers it is represented as a list with a single sub-list.
+   * @param filters Optional filters used to narrow which list item harnesses are included
+   * @return The list of items matching the given filters, grouped into sub-lists by divider.
+   */
+  async getItemsDivided(filters?: F): Promise<C[][]> {
+    const listSections = [];
+    let currentSection = [];
+    const itemsAndDividers =
+        await this.getItemsSubheadersAndDividers({item: filters, subheader: false});
+    for (const itemOrDivider of itemsAndDividers) {
+      if (itemOrDivider instanceof MatDividerHarness) {
+        listSections.push(currentSection);
+        currentSection = [];
+      } else {
+        currentSection.push(itemOrDivider);
+      }
+    }
+    listSections.push(currentSection);
+    return listSections;
+  }
+
+  /**
+   * Gets a list of harnesses representing all of the items, subheaders, and dividers
+   * (in the order they appear in the list). Use `instanceof` to check which type of harness a given
+   * item is.
+   * @param filters Optional filters used to narrow which list items, subheaders, and dividers are
+   *     included. A value of `false` for the `item`, `subheader`, or `divider` properties indicates
+   *     that the respective harness type should be omitted completely.
+   * @return The list of harnesses representing the items, subheaders, and dividers matching the
+   *     given filters.
+   */
+  getItemsSubheadersAndDividers(filters: {
```

Comment:
How about `getItemsWithSubheadersAndDividers`? The lack of commas makes the current name slightly ambiguous